x/distribution/keeper/calculation.go

```diff
@@ -0,0 +1,161 @@
+
+	// Set new historical rewards
+	k.SetValidatorHistoricalRewards(ctx, val.GetOperator(), rewards.Period, historical.Plus(current))
+
+	// Set current rewards, incrementing period by 1
+	k.SetValidatorCurrentRewards(ctx, val.GetOperator(), types.ValidatorCurrentRewards{
+		Rewards: sdk.DecCoins{},
+		Period:  rewards.Period + 1,
+	})
+
+	return rewards.Period
+}
+
+// initialize starting info for a new delegation
+func (k Keeper) initializeDelegation(ctx sdk.Context, val sdk.ValAddress, del sdk.AccAddress) {
+	// period has already been incremented
+	period := k.GetValidatorCurrentRewards(ctx, val).Period
+	validator := k.stakingKeeper.Validator(ctx, val)
+	delegation := k.stakingKeeper.Delegation(ctx, del, val)
+	// calculate delegation stake in tokens
+	// we don't store directly, so multiply delegation shares * (tokens per share)
+	stake := delegation.GetShares().Mul(validator.GetDelegatorShareExRate())
+	k.SetDelegatorStartingInfo(ctx, val, del, types.DelegatorStartingInfo{
+		PreviousPeriod: period - 1,
+		Stake:          stake,
+		Height:         uint64(ctx.BlockHeight()),
+	})
+}
+
+// calculate the rewards accrued by a delegation between two periods
+func (k Keeper) calculateDelegationRewardsBetween(ctx sdk.Context, val sdk.Validator,
+	startingPeriod, endingPeriod uint64, staking sdk.Dec) (rewards sdk.DecCoins) {
+	// sanity check
+	if startingPeriod > endingPeriod {
+		panic("startingPeriod cannot be greater than endingPeriod")
+	}
+
+	// return staking * (ending - starting)
+	starting := k.GetValidatorHistoricalRewards(ctx, val.GetOperator(), startingPeriod)
+	ending := k.GetValidatorHistoricalRewards(ctx, val.GetOperator(), endingPeriod)
+	difference := ending.Minus(starting)
+	rewards = difference.MulDec(staking)
+	return
+}
+
+// calculate the total rewards accrued by a delegation
+func (k Keeper) calculateDelegationRewards(ctx sdk.Context, val sdk.Validator, del sdk.Delegation, endingPeriod uint64) (rewards sdk.DecCoins) {
+	// fetch starting info for delegation
+	startingInfo := k.GetDelegatorStartingInfo(ctx, del.GetValidatorAddr(), del.GetDelegatorAddr())
+	startingPeriod := startingInfo.PreviousPeriod
+	stake := startingInfo.Stake
+
+	// iterate through slashes and withdraw with calculated staking for sub-intervals
+	// these offsets are dependent on *when* slashes happen - namely, in BeginBlock, after rewards are allocated...
+	// ... so we don't reduce stake for slashes which happened in the *first* block, because the delegation wouldn't have existed
+	startingHeight := startingInfo.Height + 1
+	// ... or slashes which happened in *this* block, since they would have happened after reward allocation
+	endingHeight := uint64(ctx.BlockHeight()) - 1
+	if endingHeight >= startingHeight {
+		k.IterateValidatorSlashEventsBetween(ctx, del.GetValidatorAddr(), startingHeight, endingHeight, func(height uint64, event types.ValidatorSlashEvent) (stop bool) {
```

Comment:
I'd recommend defining this function variable outside of the Iterate function

Reply:
Personally I find separate definition more confusing (assuming we don't need to reuse the definition, which we don't in this case), because it looks less like a loop - I separated the function on a new line, will that do?

Reply:
yes, is sufficient